### DISCUSSION TITLE
WiX Convert DTF Projects

### DISCUSSION THREAD

**Discussion by chrpai** [wixtoolset/issues#7155] [Questions]
Sorry this is rambling a little...

So it took me awhile to realize that Heatwave doesn't offer to upgrade DTF projects.  Eventually I found a blog article where Rob said that wix cop (took me awhile to figure out thats wix convert now .. sorry I'm mostly a VS not command line user ) doesn't upgrade your non setup code.  Well I guess to me DTF code is setup code.

As it took me awhile to realize I didn't convert my projects it means I didn't realize I still had a wix 3.11 dependency.  So then I looked at the v4 DTF projects and realized they were SDK style.

That was a bit of a suprise to me because all the native project templates in VS that I use  create .NET Framework projects with the legacy MSBuild project type and all the .NET projects using the new SDK style projects.  I didn't realize it was possible to use a legacy .NETFramework TFM in an SDK project.

Anyways I guess I'm wondering just how hard would it be for wix convert to try to support converting these like it does wixproj and would it be possible to have heatwave conditionally show the convert option only if the .csproj/.vbproj has the relevant wix targets in it.

Unless there's something I'm not thinking of, it seems like alot of the pieces are already there and it'd give a really good upgrade experience.  Otherwise I'll probably end up deleting/creating custom action projects and copying .cs files over to fill them out.

**Comment by robmen** [comment]
First, `wix convert` only processes source code. The .wixproj conversion is functionality provided by the HeatWave product.

Second, from my point of view, the largest problem is converting .cs code. Given Roslyn, it seems very possible a converter could be written to update all the namespaces and anything else that needs to be converted. Once that (large?) work item was completed, perhaps HeatWave's heuristics to upgrade .wixprojs could apply to .csprojs.

Ultimately, this is less of WiX specific issue and more of a C# issue. Sure, WiX provides the assemblies (DTF), but in the end, this is more of a question about upgrading C# project and code. Perhaps there is existing tooling in the .NET space to do exactly that.

**Comment by robmen** [comment]
PS: When I say "setup code" 99.9999999% of the time that is another way of me saying ".wxs code".

PPS: I call the code in custom actions: "custom action code". If the language of the "custom action code" matters I'll say "C++ code" or "native code"...

**Comment by chrpai** [comment]
Sorry, it's not always clear to me what is wix and what is heatwave. :)

Sticking to C#/VB CAs, the two things special a WiX csproj in my mind is the reference to DTF and the postbuild packaging.  If all other references were left alone and the TFM maintained the .cs code should compile.  Worst case would be git reset or equiv.

I can post this in the HeatWave issues for future tracking if you'd like. Maybe for a paid edition.  The real trick in my mind is knowing it's a WiX csproj.  Wouldn't want to offer upgrading just any old proj.

**Reply by robmen** [reply]
> Sorry, it's not always clear to me what is wix and what is heatwave. :)

Not a problem. I personally try to remain consistent and clear.

> the .cs code should compile. 

Do note that the namespace of DTF changed. This requires updating all of the .cs code that references DTF objects.

> I can post this in the HeatWave issues for future tracking

You are welcome to publish the feature idea.

**Reply by chrpai** [reply]
Ah, I didn't get far enough yet to notice the namespace changed.  Ok I think you sold me.

**Comment by chrpai** [comment]
I just found such a tool.  Seems I refound it as I left a comment on it a year ago.  My memory isn't what it used to be.
